installation guide - namespace per user strategy (#825)
codenvy/che-docs: src/main/pages/che-7/installation-guide/proc_configuring-namespace-strategies.adoc

[id="configuring-namespace-strategies_{context}"]
= Configuring namespace strategies

> Word 'namespace' is used interchangeably with OpenShift's 'project'.

Reviewer: ```suggestion
NOTE: The term _namespace_ (Kubernetes) is used interchangeably with _project_ (OpenShift).
```

=== Namespace per Workspace strategy

> TODO

Reviewer: Let's not put empty sections with 'TODO' into published docs. Instead, I suggest leaving out the section and filing an issue for adding it later.

=== All workspaces in one namespace strategy

> TODO

Reviewer: dtto ^

=== One Namespace per User strategy

> This strategy provides more security than ‘One namespace for all workspaces strategy’, by isolating user to his/her own namespace. At the same time it’s not so resource heavy as ‘Namespace per workspace strategy’.

Reviewer 2: " more security" - can we use something neutral. Like "more flexible", "more admin control", etc.

Author: source of this is in original issue description https://github.com/eclipse/che/issues/13488#issue-453044497
 > Use one namespace for all the workspace of every Che user (security issue)

I'm reasoning that it has better security because of isolating user. However, I can't reason how it is more flexible.
I would like to see comment from doc team, whether they feel the same that "more security" can be misinterpreted. @rkratky

Reviewer: 'more security' is indeed ambiguous.

> This strategy provides more security than ‘One namespace for all workspaces strategy’, by isolating user to his/her own namespace. At the same time it’s not so resource heavy as ‘Namespace per workspace strategy’.

Reviewer: ```suggestion

The strategy isolates each user in their own namespace.
```

> It can be configured by setting `CHE_INFRA_KUBERNETES_NAMESPACE` on Kubernetes, or respectively `CHE_INFRA_OPENSHIFT_PROJECT` on OpenShift. Value must contain one or more user placeholders. Currently supported placeholders are `<username>` and `<userid>`. By this, namespace name will be evaluated different for each user, but same for all workspaces by one user.

Reviewer: ```suggestion
To use the strategy, set the `CHE_INFRA_KUBERNETES_NAMESPACE` variable on Kubernetes or the `CHE_INFRA_OPENSHIFT_PROJECT` variable on OpenShift. The value must contain one or more user identifiers. Currently supported identifiers are `_<username>_` and `_<userid>_`.
```

> Example:

Reviewer: ```suggestion
.Example
```

> To have namespaces named like `che-ws-user1`, `che-ws-user2`, `…` set `CHE_INFRA_KUBERNETES_NAMESPACE=che-ws-<username>`.

Reviewer: ```suggestion
To assign namespace names composed of a `che-ws` prefix and individual usernames (`che-ws-user1`, `che-ws-user2`), set:

[subs="+quotes"]
----
CHE_INFRA_KUBERNETES_NAMESPACE=che-ws-__<username>__
----
```

> Limitation:

Reviewer: ```suggestion
```

> Compatible PVC strategies are `unique`, `common` and `perWorkspace`

Reviewer 3: Well, a common PVC strategy is compatible but it would mean that only 1 workspace must be run at the same time. Maybe it makes sense to clarify it here.

Reviewer 3: And do you know some incompatible namespace and PVC strategies?

Author: > Well, a common PVC strategy is compatible but it would mean that only 1 workspace must be run at the same time. Maybe it makes sense to clarify it here.

updated.

> And do you know some incompatible namespace and PVC strategies?

I don't know. TBH I'm not sure I know all namespace and all PVC strategies.

> Using One Namespace per User strategy with `common` PVC strategy will work correctly only when user runs <= 1 workspace at the same time. To enforce that, set `CHE_LIMITS_USER_WORKSPACES_RUN_COUNT=1`.

Reviewer: ```suggestion
[IMPORTANT]
====
Using the _One Namespace per User_ strategy with the `common` PVC strategy only works correctly when the user runs not more than one workspace at a time. To enforce this limit, set:

----
CHE_LIMITS_USER_WORKSPACES_RUN_COUNT=1
----
====
```
